### PR TITLE
Create gradient_norm_tracking_in_pytorch.ipynb

### PR DESCRIPTION
**Pull request by kz29**
This is the code for the new article: How to Monitor, Diagnose, and Solve Gradient Issues in Foundation Models

# Description

Include a summary of the changes and the related issue.

__Related to:__ <ClickUp/JIRA task name>

__Any expected test failures?__

---

Add a `[X]` to relevant checklist items

## ❔ This change

- [ ] adds a new feature
- [ ] fixes breaking code
- [ ] is cosmetic (refactoring/reformatting)

---

## ✔️ Pre-merge checklist

- [ ] Refactored code ([sourcery](https://sourcery.ai/))
- [ ] Tested code locally
- [ ] Precommit installed and run before pushing changes
- [ ] Added code to GitHub tests ([notebooks](workflows/test-notebooks.yml), [scripts](workflows/test-scripts.yml))
- [ ] Updated GitHub [README](../README.md)
- [ ] Updated the projects overview page on Notion

---

## 🧪 Test Configuration

- OS:
- Python version:
- Neptune version:
- Affected libraries with version:

### CODE REVIEW

File: community-code/HOW_TO_MONITOR_DIAGNOSE_AND_SOLVE_GRADIENT_ISSUES_IN_FOUNDATION_MODELS/gradient_norm_tracking_in_pytorch .ipynb [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    "        # Log gradient norms\n",
+    "        log_gradient_norms(model, step + epoch * len(train_dataloader))\n",
+    "\n",
+    "        optimizer.step()\n",
+    "\n",
+    "        # Log Loss to Neptune Scale\n",
+    "        run.log_metrics({\"loss\": loss.item()}, step=step + epoch * len(train_dataloader))\n",
+    "\n",
+    "# Add tags and close the run\n",
+    "run.add_tags([\"gradient_tracking\", \"pytorch\", \"transformers\"])\n",
```
[the remark this answers is not in the document]
Comment (kz29) [reply community-code/HOW_TO_MONITOR_DIAGNOSE_AND_SOLVE_GRADIENT_ISSUES_IN_FOUNDATION_MODELS/gradient_norm_tracking_in_pytorch .ipynb:255]: Thank you!

File: community-code/HOW_TO_MONITOR_DIAGNOSE_AND_SOLVE_GRADIENT_ISSUES_IN_FOUNDATION_MODELS/gradient_norm_tracking_in_pytorch .ipynb [hunk 2]
```diff
@@ -0,0 +1,281 @@
+{
```
[the remark this answers is not in the document]
Comment (kz29) [reply community-code/HOW_TO_MONITOR_DIAGNOSE_AND_SOLVE_GRADIENT_ISSUES_IN_FOUNDATION_MODELS/gradient_norm_tracking_in_pytorch .ipynb]: I updated the code! Thank you for pointing it out!

**Comment by SiddhantSadangi** [comment]
@sourcery-ai review

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    "run = Run(\n",
+    "    api_token=\"YOUR_API_TOKEN\",  # replace with your Neptune API token\n",
+    "    project=\"YOUR_WORKSPACE/YOUR_PROJECT\",  # replace with your workspace and project name\n",
+    "    experiment_name=\"gradient_tracking_lr=1\",\n",
+    ")\n",
+    "\n",
+    "run.log_configs(\n",
+    "    {\n",
+    "        \"learning_rate\": 1,\n",
+    "        \"batch_size\": 1,\n",
```
[the remark this answers is not in the document]
Comment (SiddhantSadangi) [reply community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:145]: @kz29 - Could you address this inconsistency? Ideally, we'd want a single configs/params dictionary that is reused in the code and logged to the run. Any change in the parameters can then be made in this one place, instead of multiple places in the code

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "optimizer = optim.Adam(model.parameters(), lr=1)\n",
+    "\n",
+    "model.train()\n",
+    "for epoch in range(10):\n",
+    "    for step, batch in enumerate(train_dataloader):\n",
+    "        inputs = {k: v.to(device) for k, v in batch.items() if k in tokenizer.model_input_names}\n",
+    "        labels = batch[\"labels\"].to(device)\n",
```
[the remark this answers is not in the document]
Comment (SiddhantSadangi) [reply community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:212]: @kz29 - should be fixed

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/.ipynb_checkpoints/gradient_norm_tracking_in_pytorch -checkpoint.ipynb
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/.ipynb_checkpoints/gradient_norm_tracking_in_pytorch -checkpoint.ipynb:1]: `*checkpoint.ipynb` files should not be committed

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/README.md
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/README.md:1]: Since this blog only contains a notebook, the installation steps can be moved to the notebook itself and the README deleted

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    "train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=8, collate_fn=data_collator)\n",
+    "\n",
+    "tokenized_datasets = dataset.map(tokenize_function, batched=True)\n",
+    "tokenized_datasets.set_format(\n",
+    "    type=\"torch\", columns=[\"input_ids\", \"attention_mask\", \"token_type_ids\", \"label\"]\n",
+    ")\n",
+    "\n",
+    "# model = BertForSequenceClassification.from_pretrained('bert-base-uncased')\n",
+    "# # Move model to device\n",
+    "# model.to(device)\n",
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:97]: Instead of commenting here, this can be moved to after BertConfig() so that the statement will be relevant irrespective of using a pretrained model or not

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    "    hidden_size=768,\n",
+    "    intermediate_size=3072,\n",
+    "    layer_norm_eps=1e-12,\n",
+    "    output_attentions=False,\n",
+    "    output_hidden_states=False,\n",
+    "    num_labels=2,\n",
+    ")\n",
+    "\n",
+    "\n",
+    "# config = BertConfig(hidden_dropout_prob=0.5)  # Add more dropout\n",
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:116]: Remove if not needed

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "run = Run(\n",
+    "    api_token=\"YOUR_API_TOKEN\",  # replace with your Neptune API token\n",
+    "    project=\"YOUR_WORKSPACE/YOUR_PROJECT\",  # replace with your workspace and project name\n",
+    "    experiment_name=\"gradient_tracking_lr=1\",\n",
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:139]: If using a global `configs` dict, the name can be parameterized based on the value of `lr` in the configs.

```suggestion
    "    experiment_name=f\"gradient_tracking_lr={configs.lr}\"\n",
```

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    ")\n",
+    "\n",
+    "run.log_configs(\n",
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:142]: If using a global `configs` dict, you can just pass the dict to `log_configs()`.

```suggestion
    "run.log_configs(configs)\n",
```

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    "        return  # Skip logging for this step\n",
+    "\n",
+    "    with torch.no_grad():  # Prevent building a computation graph during norm computation\n",
+    "        for name, param in model.named_parameters():\n",
+    "            if param.grad is not None:\n",
+    "                # Optional: skip small/irrelevant layers if needed\n",
+    "                # if not name.startswith(\"encoder.layer.\"): continue\n",
+    "\n",
+    "                grad_norm = param.grad.norm().item()\n",
+    "                run.log_metrics({f\"gradients/{name}\": grad_norm}, step=step)"
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:188]: Instead of a separate logging call for each grad, it'ld be more efficient to collect all grads and log them using a single call.
_Please check the suggestion before applying. I've winged it in the raw text editor :)_  

```suggestion
    "    grad_norm = {}\n",
    "    with torch.no_grad():  # Prevent building a computation graph during norm computation\n",
    "        for name, param in model.named_parameters():\n",
    "            if param.grad is not None:\n",
    "                # Optional: skip small/irrelevant layers if needed\n",
    "                # if not name.startswith(\"encoder.layer.\"): continue\n",
    "                \n",
    "                grad_norm[f\"gradients/{name}\"] = param.grad.norm().item()\n",
    "    run.log_metrics(grad_norm, step=step)\n"
```

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb (hunk 4 above)
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:212]: I haven't tested it, but it might be faster to move the inputs and labels to the relevant device outside of the loop.

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# Core ML libraries
+torch==1.10.2+cu115  # CUDA 11.5-compatible version
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/requirements.txt:2]: Since users might have a different version of CUDA installed, this should ideally be parameterized in the notebook, or they should be informed about installing a version compatible with their CUDA version

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+transformers==4.38.2
+datasets==3.5.0
+
+# Optional: Logging, tracking, utilities
+neptune==1.15.0
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/requirements.txt:7]: `neptune` is deprecated and isn't needed anymore

```suggestion
```

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+neptune-scale
+tqdm
+numpy
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/requirements.txt:10]: Please check if the code works for `numpy>=2.0`. If not, please constrain to `numpy<2.0` here

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]
Thanks for the blog, @kz29 , and for your patience here.

I've left a few comments and suggestions.

On a high-level, I'd prefer using a global configs dictionary so that parameters can be read from it and logged to Neptune without being hardcoded in multiple places in the code.

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb (hunk 5 above)
Comment (kz29) [reply community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb]: Thank you, @SiddhantSadangi, for the suggestions. I updated the notebook accordingly.

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+   "outputs": [],
+   "source": [
+    "run = Run(\n",
+    "    api_token=configs[\"api_token\"],\n",
+    "    project=configs[\"project\"],\n",
+    "    experiment_name=f\"gradient_tracking_lr={configs['learning_rate']}\",\n",
+    ")\n",
+    "\n",
+    "# Log configs to Neptune\n",
+    "run[\"parameters\"] = configs\n",
```
Comment (SiddhantSadangi) [inline community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb:169]: The method to log such configs is [`log_configs()`](https://docs.neptune.ai/log_configs/).

In this case, we will need to prefix `parameters/` to the config keys so that they are all logged in the `parameters` namespace

File: community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/.ipynb_checkpoints/gradient_norm_tracking_in_pytorch -checkpoint.ipynb
Comment (SiddhantSadangi) [reply community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/.ipynb_checkpoints/gradient_norm_tracking_in_pytorch -checkpoint.ipynb:1]: @kz29 - Looks like it has not been deleted yet. You'll need to add `.ipynb_checkpoints` to `.gitignore`

**Review by SiddhantSadangi — APPROVED** [review]
